R/dependecies.R

```diff
@@ -0,0 +1,437 @@
+        obj2$download_path != obj1$download_path) {
+        unlink(obj2$download_path, TRUE, TRUE)
+      }})
+
+  ## Swap a package from list1 into list2
+  list2_swap <- vapply(names1[swap_package2for1]
+    , function(n) which(names2 == n)
+    , integer(1))
+
+  list2[list2_swap] <- list1[swap_package2for1]
+  list2
+}
+
+#' Either use the current lockbox library DESCRIPTION
+#' file or download the accurate remote DESCRIPTION file.
+get_dependencies <- function(package, lock) {
+  locked_package <- package
+
+  ## When we have a dependency package that has a version in lockbox
+  ## we will substitute that version for our package version if that package
+  ## version is missing (meaning use any version available) or this required
+  ## version is less than that already in the lockbox
+  if (locked_package$is_dependency_package &&
+    !is.null(locked_package$latest_version_in_lockbox) &&
+    (is.na(locked_package$version) ||
+      package_version(as.character(locked_package$version)) <
+      package_version(as.character(locked_package$latest_version_in_lockbox)))) {
+      locked_package$version <- locked_package$latest_version_in_lockbox
+  }
+  if (!is.na(locked_package$version) && exists_in_lockbox(locked_package)) {
+    dependencies <- dependencies_from_description(locked_package
+      , description_file_for(locked_package$name
+        , dirname(lockbox_package_path(locked_package))))
+  } else {
+    cat(crayon::blue("."))
+    output <- tryCatch(get_remote_dependencies(package), error = function(e) e)
+    if (is(output, "error")) {
+      stop(crayon::red(paste0("Dependencies could not be resolved for package: "
+        , package$name, " version: ", package$version
+        , " remote:", package$remote)))
+    } else {
+      package <- output$package
+      dependencies <- output$dependencies
+    }
+  }
+  dependencies <- strip_available_dependencies(dependencies)
+  dependencies <- strip_duplicate_dependencies(dependencies)
+  dependencies <- lapply(dependencies, add_latest_version_in_lockbox)
+  dependencies <- lapply(dependencies, function(dep) {
+    dep$parent_package <- package$name
+    dep})
+  dependencies <- lapply(dependencies, replace_with_lock, lock)
+  list(package = package, dependencies = dependencies)
+}
+
+strip_available_dependencies <- function(dependencies) {
+  dependencies <- lapply(dependencies
+    , function(package) {
+      if(package$name %in% pesky_namespaces) NULL
+      else package
+    })
+ dependencies[!vapply(dependencies, is.null, logical(1))]
+}
+
+#' We don't trust package authors to only put a package in once
+strip_duplicate_dependencies <- function(dependencies) {
+  dependencies[!duplicated(vapply(dependencies, function(d) d$name
+    , character(1)))]
+}
+
+#' Get the dependencies for a given package
+get_remote_dependencies <- function(package) {
+  UseMethod("get_remote_dependencies")
+}
+
+#' If a package is local we just read from the directory given
```

Comment:
You need to stop using literate documentation if you're not actually going to comment the params and return value :)